onnxruntime/core/external_ops/pyop.cc

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef _DEBUG
+#undef _DEBUG
+#include <Python.h>
+#define _DEBUG
+#else
+#include <Python.h>
+#endif
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include "numpy/arrayobject.h"
+#include <functional>
+#include <iostream>
+#include <sstream>
+#include <numeric>
+#include <vector>
+#include <mutex>
+#include <functional>
+#include <unordered_map>
+
+using namespace std;
+namespace PythonFuncionWrapper {
+
+#ifdef _WIN32
+#define PYOP_EXPORT extern "C" __declspec(dllexport)
+#else
+#define PYOP_EXPORT extern "C"
+#endif
+
+struct Finalizer
+{
+    ~Finalizer() {
+        Py_Finalize();
+    }
+};
+
+class Scope
+{
+public:
+    Scope(const vector<PyObject*> objs = {}): objs_(objs) {
+        mtx_.lock();
+    }
+    ~Scope() {
+        for (auto obj: objs_) {
+            Py_XDECREF(obj);
+        }
+        mtx_.unlock();
+    }
+    void Add(PyObject* obj) {
+        objs_.push_back(obj);
+    }
+private:
+    static std::mutex mtx_;
+    vector<PyObject*> objs_;
+};
+
+std::mutex Scope::mtx_;
+
+PYOP_EXPORT bool Initialize() 
+{
+    Scope scope;
+    Py_Initialize();
+    if (_import_array() < 0) {
+        return false;
+    }
+    auto path_list = PySys_GetObject("path");// do not release it
+    if (nullptr == path_list || !PyList_Check(path_list) ||
+        PyList_Append(path_list, PyUnicode_FromString(".")) != 0) {
+        return false; 
+    }
+    static Finalizer finalizer;
+    return true;
+}
+
+PYOP_EXPORT const char* GetLastErrorMessage(std::string& err)
+{
+    Scope scope;
+    if (PyErr_Occurred()) {
+        PyObject *type, *value, *trace;
+        PyErr_Fetch(&type, &value, &trace);
+        if (nullptr != value) {
+            auto pyVal = PyObject_Repr(value);
+            scope.Add(pyVal);
+            auto pyStr = PyUnicode_AsEncodedString(pyVal, "utf-8", "Error ~");
+            scope.Add(pyStr);
+            err = PyBytes_AS_STRING(pyStr);
+        }
+        PyErr_Restore(type, value, trace);
+    }
+    return err.c_str();
+}
+
+PyObject* MakePyObj (const void* data, int32_t type, const vector<int64_t>& dim)
+{
+    std::vector<npy_intp> np_dim;
+    for (auto d: dim) {
+        np_dim.push_back(static_cast<npy_intp>(d));
+    }
+    auto pyObj = static_cast<PyObject*>(PyArray_EMPTY(static_cast<int>(np_dim.size()), np_dim.data(), type, 0));
```

Comment:
Do we need Something like PyArray_INCREF? The documentation is not really clear about it.